todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/taskdetail/TaskDetailViewModel.java

```diff
@@ -0,0 +1,198 @@
+            final String taskId = deleteTaskIntent.taskId();
+            return TaskDetailAction.DeleteTask.create(taskId);
+        }
+
+        if (intent instanceof TaskDetailIntent.CompleteTaskIntent) {
+            TaskDetailIntent.CompleteTaskIntent completeTaskIntent = (TaskDetailIntent.CompleteTaskIntent) intent;
+            final String taskId = completeTaskIntent.taskId();
+            return TaskDetailAction.CompleteTask.create(taskId);
+        }
+
+        if (intent instanceof TaskDetailIntent.ActivateTaskIntent) {
+            TaskDetailIntent.ActivateTaskIntent activateTaskIntent = (TaskDetailIntent.ActivateTaskIntent) intent;
+            final String taskId = activateTaskIntent.taskId();
+            return TaskDetailAction.ActivateTask.create(taskId);
+        }
+
+        if (intent instanceof TaskDetailIntent.GetLastState) {
+            return TaskDetailAction.GetLastState.create();
+        }
+        throw new IllegalArgumentException("do not know how to treat this intent " + intent);
+    }
+
+    private static BiFunction<TaskDetailViewState, TaskDetailResult, TaskDetailViewState> reducer =
+            (previousState, result) -> {
+                TaskDetailViewState.Builder stateBuilder = previousState.buildWith();
+                if (result instanceof TaskDetailResult.GetLastState) {
+                    return stateBuilder.build();
+                }
+                if (result instanceof TaskDetailResult.PopulateTask) {
+                    TaskDetailResult.PopulateTask populateTaskResult =
+                            (TaskDetailResult.PopulateTask) result;
+                    switch (populateTaskResult.status()) {
+                        case SUCCESS:
+                            Task task = checkNotNull(populateTaskResult.task());
+                            stateBuilder.title(task.getTitle());
+                            stateBuilder.description(task.getDescription());
+                            stateBuilder.active(task.isActive());
+                            return stateBuilder.build();
+                        case FAILURE:
+                            Throwable error = checkNotNull(populateTaskResult.error());
+                            return stateBuilder.error(error).build();
+                        case IN_FLIGHT:
+                            // nothing to do
+                            return stateBuilder.build();
+                    }
+                }
+                if (result instanceof TaskDetailResult.DeleteTaskResult) {
+                    TaskDetailResult.DeleteTaskResult deleteTaskResult =
+                            (TaskDetailResult.DeleteTaskResult) result;
+                    switch (deleteTaskResult.status()) {
+                        case SUCCESS:
+                            return stateBuilder.taskDeleted(true).build();
+                        case FAILURE:
+                            return stateBuilder.taskComplete(false).error(deleteTaskResult.error()).build();
```

Comment:
typo?
Does `taskComplete` need to change?

Reply:
ouch, too much code! 💃